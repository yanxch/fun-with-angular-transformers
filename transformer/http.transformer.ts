import * as ts from 'typescript';
import {AngularCompilerPlugin} from '@ngtools/webpack';
import axios, { AxiosRequestConfig, AxiosPromise } from 'axios';

// Build with:
// Terminal 1: tsc --skipLibCheck --module umd -w
// Terminal 2: ng build --aot --plugin ~dist/out-tsc/plugins.js
// Terminal 3: ng build --plugin ~dist/out-tsc/plugins.js
// ng serve --plugin ~dist/out-tsc/plugins.js


// 1. CallExpression
// 2. CallExpression.expression === PropertyAccessExpression
// 3. PropertyAccessExpression.expression === PropertyAccessExpression
// 4. PropertyAccessExpression.name === 'http'

// Get the HTTP Method from the name of the first PropertyAccessExpression
// Get the Arguments of the CallExpression (Array mit Stringliteral == URL)
// Create Observable with Observable.of({}) an replace the CallExpression with it


function isHttpExpression(node: ts.Node): node is ts.CallExpression {
  return ts.isCallExpression(node) &&
    ts.isPropertyAccessExpression(node.expression) &&
    ts.isPropertyAccessExpression(node.expression.expression) &&
    node.expression.expression.name.text === 'http';
}

export function httpTransformerFactory(acp: AngularCompilerPlugin) {
  return (context: ts.TransformationContext) => {

    return (rootNode: ts.SourceFile) => {
      console.log('visit: ' + rootNode.fileName);

      function visit(node: ts.Node): ts.Node {

        if (isHttpExpression(node)) {
          const httpMethodExpression = (node.expression as ts.PropertyAccessExpression);
          const httpExpression = (httpMethodExpression.expression as ts.PropertyAccessExpression);
          const httpUrl = node.arguments[0] as ts.StringLiteral;
          console.log('1. PropertyAccessExpression - Name: ' +  httpMethodExpression.name.text);
          console.log('2. PropertyAccessExpression - Name: ' +  httpExpression.name.text);
          console.log('3. URL: ' +  httpUrl.text);

          console.log('Found a http call');

          axios
            .get(httpUrl.text)
            .then(function (response) {
              // handle success
              const dataString = JSON.stringify(response.data);

              const callExpression = ts.createCall(
                ts.createIdentifier('of'),
                undefined,
                [ts.createStringLiteral(dataString)]
              );

              console.log(callExpression);
            });
        }

        return ts.visitEachChild(node, visit, context);
      }

      return ts.visitNode(rootNode, visit);
    };
  };
}


/*function isSubscribeExpression(node: ts.Node, checker: ts.TypeChecker): node is ts.CallExpression {
  // ts.isBinaryExpression
  // ts.isCallExpression
  // ts.isClassDeclaration
  // ts.is

  return ts.isCallExpression(node) &&
    node.parent && ts.isExpressionStatement(node.parent) &&
    ts.isPropertyAccessExpression(node.expression) &&
    node.expression.name.text === 'subscribe' &&
    rxjsTypes.includes(getTypeAsString(node, checker));
} */