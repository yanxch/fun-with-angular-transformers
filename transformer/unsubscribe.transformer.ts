import * as ts from 'typescript';
import {AngularCompilerPlugin} from '@ngtools/webpack';

// Build with:
// Terminal 1: tsc --skipLibCheck --module umd -w
// Terminal 2: ng build --aot --plugin ~dist/out-tsc/plugins.js

const rxjsTypes = [
  'Observable',
  'BehaviorSubject',
  'Subject',
  'ReplaySubject'
];
/**
 * looking into:
 *    - call expressions within a
 *    - expression statement only
 *    - that wraps another call expression where a property is called with subscribe 
 *    - and the type is contained in rxjsTypes
 * 
 * Attention: If there is another method call to "subscribe" anywhere, that is not coming from RxJs, we are screwed!
 */
function isSubscribeExpression(node: ts.Node, checker: ts.TypeChecker): node is ts.CallExpression {
  return ts.isCallExpression(node)  
    && node.parent && ts.isExpressionStatement(node.parent) 
    && ts.isPropertyAccessExpression(node.expression)
    && node.expression.name.text === 'subscribe'
    && rxjsTypes.includes(getTypeAsString(node, checker));
} 

function getTypeAsString(node: ts.CallExpression, checker: ts.TypeChecker) {
  const type = checker.getTypeAtLocation((node.expression as ts.PropertyAccessExpression | ts.CallExpression).expression);
  console.log('TYPE: ', type.symbol.name);
  return type.symbol.name;
}

/**
 * Takes a subscibe call expression and wraps it with:
 * this.subscriptions.push(node)
 */
function wrapSubscribe(node: ts.CallExpression) {
  
  // console.log(node.expression);

  return ts.createCall(
    ts.createPropertyAccess(
      ts.createPropertyAccess(ts.createThis(), 'subscriptions'),
      'push'
    ),
    undefined,
    [node]
  );
}

function logClassFound(node: ts.ClassDeclaration) {
  console.log('Found class: ', node.name.escapedText);
}

function isComponent(node: ts.ClassDeclaration) {
  return node.decorators && node.decorators.filter(d => d.getFullText().trim().startsWith('@Component')).length > 0;
}

/**
 * creates an empty array property:
 * subscriptions = [];
 */
function createSubscriptionsArray() {
  return ts.createProperty(
    undefined, 
    undefined, 
    'subscriptions', 
    undefined, 
    undefined, 
    ts.createArrayLiteral()
  );
}

function isNgOnDestroyMethod(node: ts.ClassElement): node is ts.MethodDeclaration {
  return ts.isMethodDeclaration(node) && (node.name as ts.Identifier).text == 'ngOnDestroy';
}

function hasNgOnDestroyMethod(node: ts.ClassDeclaration) {
  return node.members.filter(node => isNgOnDestroyMethod(node)).length > 0;
}

function getNgOnDestroyMethod(node: ts.ClassDeclaration) {
  const n = node.members
    .filter(node => isNgOnDestroyMethod(node))
    .map(node => node as ts.MethodDeclaration);
   return n[0];
}

function createNgOnDestroyMethod() {
  return ts.createMethod(
    undefined,
    undefined,
    undefined,
    'ngOnDestroy',
    undefined,
    [],
    [],
    undefined,
    ts.createBlock([], true)
  );
}

function createUnsubscribeStatement() {
  return ts.createExpressionStatement(
    ts.createCall(
      ts.createPropertyAccess(
        ts.createPropertyAccess(ts.createThis(), 'subscriptions'),
        'forEach'
      ),
      undefined,
      [
        ts.createArrowFunction(
          undefined,
          undefined,
          [
            ts.createParameter(undefined, undefined, undefined, 'sub', undefined, undefined, undefined)
          ],
          undefined,
          ts.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
          ts.createCall(
            ts.createPropertyAccess(ts.createIdentifier('sub'), 'unsubscribe'),
            undefined,
            []
          )
        )
      ]
    )
  );
}

export function unsubscribeTransformerFactory(acp: AngularCompilerPlugin) {
  return (context: ts.TransformationContext) => {

    const checker = acp.typeChecker;

    return (rootNode: ts.SourceFile) => {
   
      function visit(node: ts.Node): ts.Node {
  
        if (isSubscribeExpression(node, checker)) {
          return wrapSubscribe(node);
        }
        
        if (ts.isClassDeclaration(node)) {
          logClassFound(node);
  
          if (isComponent(node)) {
            // 1.
            node.members = ts.createNodeArray([...node.members, createSubscriptionsArray()]);
  
            // 2.
            if (!hasNgOnDestroyMethod(node)) {
              node.members = ts.createNodeArray([...node.members, createNgOnDestroyMethod()]);
            }
  
            // 3.
            const ngOnDestroyMethod = getNgOnDestroyMethod(node);
            ngOnDestroyMethod.body.statements = ts.createNodeArray([...ngOnDestroyMethod.body.statements, createUnsubscribeStatement()]);
          }  
        } 
        
        return ts.visitEachChild(node, visit, context);
      }
  
      return ts.visitNode(rootNode, visit);
    };
  };
}
