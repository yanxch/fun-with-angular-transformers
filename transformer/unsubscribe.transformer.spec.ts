import * as ts from 'typescript';

import {MockAotContext, MockCompilerHost} from './mocks';
import {unsubscribeTransformerFactory} from './unsubscribe.transformer';

describe('unsubscribe transformer', () => {
  it('should create ngOnDestroy method and subscriptions field', () => {
    const enhanced = convert(`
      @Component({
        selector: 'test-selector'
      }) 
      class Foo {

      }
    `);
    expect(enhanced).toMatchSnapshot();
  });

  it('should keep existing ngOnDestroy method', () => {
    const enhanced = convert(`
      @Component({
        selector: 'test-selector'
      }) 
      class Foo {
        ngOnDestroy() {
          console.log('Keep me there plz!');
        }
      }
    `);
    expect(enhanced).toMatchSnapshot();
  });
});

function convert(source: string) {
  
  const baseFileName = 'someFile';
  const moduleName = '/' + baseFileName;
  const fileName = moduleName + '.ts';
  const context = new MockAotContext('/', {[baseFileName + '.ts']: source});
  const host = new MockCompilerHost(context);

  const sourceFile =
      ts.createSourceFile(fileName, source, ts.ScriptTarget.Latest, /* setParentNodes */ true);
  const program = ts.createProgram(
      [fileName], {
        module: ts.ModuleKind.CommonJS,
        target: ts.ScriptTarget.ES2017,
      },
      host);
  const moduleSourceFile = program.getSourceFile(fileName);
  const typeChecker = program.getTypeChecker();

  const transformers: ts.CustomTransformers = {
    before: [unsubscribeTransformerFactory({ typeChecker } as any)]
  };
  let result = '';
  const emitResult = program.emit(
      moduleSourceFile, (emittedFileName, data, writeByteOrderMark, onError, sourceFiles) => {
        if (fileName.startsWith(moduleName)) {
          result = data;
        }
      }, undefined, undefined, transformers);
  return result;
  
}
