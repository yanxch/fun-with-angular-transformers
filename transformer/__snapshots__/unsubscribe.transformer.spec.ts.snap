// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unsubscribe transformer should create ngOnDestroy method and subscriptions field 1`] = `
"var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === \\"object\\" && typeof Reflect.decorate === \\"function\\") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
let Foo = class Foo {
    constructor() {
        this.subscriptions = [];
    }
    ngOnDestroy() {
        this.subscriptions.forEach(sub => sub.unsubscribe());
    }
};
Foo = __decorate([
    Component({
        selector: 'test-selector'
    })
], Foo);
"
`;

exports[`unsubscribe transformer should keep existing ngOnDestroy method 1`] = `
"var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === \\"object\\" && typeof Reflect.decorate === \\"function\\") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
let Foo = class Foo {
    constructor() {
        this.subscriptions = [];
    }
    ngOnDestroy() {
        console.log('Keep me there plz!');
        this.subscriptions.forEach(sub => sub.unsubscribe());
    }
};
Foo = __decorate([
    Component({
        selector: 'test-selector'
    })
], Foo);
"
`;
